onnxruntime/test/providers/provider_test_utils.cc

```diff
@@ -769,7 +769,35 @@ void OpTester::Run(
 
       for (auto& entry : *execution_providers) {
         provider_types += entry->Type() + ":";
-        ASSERT_STATUS_OK(session_object.RegisterExecutionProvider(std::move(entry)));
+
+        std::unique_ptr<IExecutionProvider> execution_provider;
+        if (entry->Type() == onnxruntime::kCpuExecutionProvider)
+          execution_provider = DefaultCpuExecutionProvider();
+        else if (entry->Type() == onnxruntime::kCudaExecutionProvider)
+          execution_provider = DefaultCudaExecutionProvider();
+        else if (entry->Type() == onnxruntime::kDnnlExecutionProvider)
+          execution_provider = DefaultDnnlExecutionProvider();
+        else if (entry->Type() == onnxruntime::kOpenVINOExecutionProvider)
+          execution_provider = DefaultOpenVINOExecutionProvider();
+        else if (entry->Type() == onnxruntime::kNupharExecutionProvider)
+          execution_provider = DefaultNupharExecutionProvider();
+        else if (entry->Type() == onnxruntime::kTensorrtExecutionProvider)
+          execution_provider = DefaultTensorrtExecutionProvider();
+        else if (entry->Type() == onnxruntime::kNnapiExecutionProvider)
+          execution_provider = DefaultNnapiExecutionProvider();
+        else if (entry->Type() == onnxruntime::kRknpuExecutionProvider)
+          execution_provider = DefaultRknpuExecutionProvider();
+        else if (entry->Type() == onnxruntime::kAclExecutionProvider)
+          execution_provider = DefaultAclExecutionProvider();
+        else if (entry->Type() == onnxruntime::kArmNNExecutionProvider)
+          execution_provider = DefaultArmNNExecutionProvider();
+        else if (entry->Type() == onnxruntime::kRocmExecutionProvider)
+          execution_provider = DefaultRocmExecutionProvider();
+        // skip if execution provider is disabled
+        if (execution_provider == nullptr)
+          continue;
+
+        ASSERT_STATUS_OK(session_object.RegisterExecutionProvider(std::move(execution_provider)));
```

Comment:
Why is this change necessary? It looks like it changes the meaning of the `execution_providers` argument, by switching the (potentially custom) passed-in object to a default instance of the same type.

Reply:
The gradient unit tests use the Jacobian matrix to validate the gradient function.  The tests end up calling the OpTester::Run function multiple times.

Before adding the dnnl execution provider to the gradient tests the code always executed the else branch when running the gradient tests. Avoiding this block of code.  The problem is that the `entry->Type()` is moved into the `session_object` the first time it is called.  The means the next time the Run function is called the memory found in `entry->Type()` has been moved resulting in an error.

This change creates a new execution provider based on the value found in entry->Type() and then moves the new execution provider in to the `session_object.RegisterExecutionProvider( ... )` leaving the `entry->Type()` unchanged so it can be referenced for the next time the `Run` function is called.

Reply:
Hmm, the problem here is that `entry->Type()` isn't sufficient to reconstruct all execution providers. For example, a `CudaExecutionProvider` might be constructed with a [`CudaExecutionProviderInfo`](https://github.com/microsoft/onnxruntime/blob/64709b1335380259b6b6e40cd02f5c3aeee7b38d/onnxruntime/core/providers/cuda/cuda_execution_provider_info.h) that is different from the default, and this change would silently drop those settings.

I think I understand the problem though... the documentation for the `execution_provider` argument isn't clear about the fact that the passed-in providers are consumed by the method (and moved into the session). Is it possible to fix this by recreating the dnnl EP each time this method is called?

Or does the problem arise because the gradient tests make  multiple calls to `OpTester::Run()` internally? :(. In that case, perhaps we should pass an optional vector of factory functions (`std::vector<std::function<std::unique_ptr<IExecutionProvider>>>*` ) that can recreate the EP(s) as necessary?

Reply:
> Is it possible to fix this by recreating the dnnl EP each time this method is called?
>
> Or does the problem arise because the gradient tests make multiple calls to OpTester::Run() internally?

None of the code written by me calls the `OpTester::Run() ` directly it is done indirectly somewhere in the test setup. Because of that, I don't think I could recreate the dnnl EP for each method call.

Reply:
OK, I've thought about this some more, and I think the root of the problem is that `GradientOpTester::Run()` (for some unknown reason) has an unused `execution_providers` argument, and attempting to use it causes problems because the execution providers end up being consumed multiple times. This part of the code looks like it needs to be refactored, but I wouldn't want that to derail this PR.

Given that, is it feasible to:
1. Revert the change to `OpTester::Run()` to avoid the breaking change, and
2. Localize the recreate-based-on-type logic to `GradientOpTester::Run()`?

If the tests don't pass with this suggestion, I'd prefer to disable the failing tests with a TODO/file an issue, rather than potentially break other callers of `OpTester::Run()`.

Reply:
I reverted the change and all of the tests completed without the segfault error that was the original reason for making the change. So I was able to remove the lines in question and all the tests pass.

Yay, we still have the test coverage without introducing the issue you were concerned with.